share/regulate/regulator.py

```diff
@@ -0,0 +1,119 @@
+from django.conf import settings
+
+from share import exceptions
+from share.models import RegulatorLog
+from share.util.extensions import Extensions
+
+
+class RegulatorConfigError(exceptions.ShareException):
+    pass
+
+
+class InfiniteRegulationError(exceptions.ShareException):
+    pass
+
+
+class Regulator:
+    VERSION = 1
+
+    def __init__(
+            self,
+            ingest_job=None,
+            source_config=None,
+            regulator_config=None,
+            validate=True,
+    ):
+        assert not ingest_job or not source_config, 'Provider ingest_job or source_config, not both'
+
+        self.job = ingest_job
+        self._logs = []
+
+        if ingest_job and not source_config:
+            source_config = ingest_job.suid.source_config
+
+        self._custom_steps = Steps(
+            self,
+            source_config.regulator_steps if source_config else None,
+            validate=validate,
+        )
+        self._default_steps = Steps(
+            self,
+            regulator_config or settings.SHARE_REGULATOR_CONFIG,
+            validate=validate
+        )
+
+    def regulate(self, graph):
+        try:
+            self._custom_steps.run(graph)
```

Comment:
Are there any situations where you'd need to run steps after the default steps?

Reply:
Not that I've run into.

I think of custom steps as a place for source-specific logic that makes more sense expressed imperatively. Right now that's what the `RunPython` chain links are for, but moving it here lets the transformers can remain declarative. The transformers just allocate information into buckets, then source-specific regulator steps can do additional processing to get it ready for actual regulation. I think it helps to guarantee the transformer's output goes directly into custom regulation, without anything in between.